File: easy/101-200/173/rust/src/main.rs
// Copyright 2022 CJ Harries
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use lazy_static::lazy_static;
use regex::Regex;
use std::collections::HashMap;

lazy_static! {
    static ref CONVERSION_PATTERN: Regex =
        Regex::new(r"\s*(?P<amount>[^ ]*) (?P<from>.*) to (?P<to>.*)\s*").unwrap();
    static ref CONVERSION_MAP: HashMap<String, HashMap<String, f64>> = {
        let mut map = HashMap::new();
        map.insert(
            "inches".to_string(),
            HashMap::from_iter([
                ("metres".to_string(), 1.0 / 39.37),
                ("miles".to_string(), 1.0 / 63360.0),
                ("attoparsecs".to_string(), 1.0 / 1.215),
            ]),
        );
        map.insert(
            "metres".to_string(),
            HashMap::from_iter([
                ("inches".to_string(), 39.37),
                ("miles".to_string(), 1.0 / 1609.0),
                ("attoparsecs".to_string(), 32.408),
            ]),
        );
        map.insert(
            "miles".to_string(),
            HashMap::from_iter([
                ("inches".to_string(), 63360.0),
                ("metres".to_string(), 1609.0),
                ("attoparsecs".to_string(), 52155.0),
            ]),
        );
        map.insert(
            "attoparsecs".to_string(),
            HashMap::from_iter([
                ("inches".to_string(), 1.0 / 1.215),
                ("metres".to_string(), 1.0 / 32.408),
                ("miles".to_string(), 1.0 / 52155.0),
            ]),
        );
        map.insert(
            "kilograms".to_string(),
            HashMap::from_iter([
                ("pounds".to_string(), 0.45359237),
                ("ounces".to_string(), 28.349523125),
            ]),
        );
        map
    };
}

#[cfg(not(tarpaulin_include))]
fn main() {
    println!("rad");
}

fn convert_units(input: &str) -> String {
    let captures = CONVERSION_PATTERN.captures(input).unwrap();
    let amount = captures
        .name("amount")
        .unwrap()
        .as_str()
        .parse::<f64>()
        .unwrap();
    let from = captures.name("from").unwrap().as_str().to_string();
    let to = captures.name("to").unwrap().as_str().to_string();
    match CONVERSION_MAP.get(&from) {
        Some(map) => match map.get(&to) {
            Some(conversion) => format!(
                "{} {} is {} {}",
                amount,
                from,
                (conversion * amount * 100000.0).round() / 100000.0,
                to
            ),
            None => format!("{} {} cannot be converted to {}", amount, from, to),
        },
        None => "error".to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_convert_units() {
        assert_eq!(
            "3.000000 metres is 118.110000 inches",
            convert_units("3 metres to inches")
        );
        assert_eq!(
            "3.000000 metres cannot be converted to pounds",
            convert_units("3 metres to pounds")
        );
    }
}

File: easy/101-200/173/go/main.go
// Copyright 2021 CJ Harries
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"fmt"
	"regexp"
	"strconv"
)

var unitConversions = map[string]map[string]float64{
	"metres": {
		"inches":      39.37,
		"miles":       1 / 1609,
		"attoparsecs": 32.408,
	},
	"inches": {
		"metres":      1 / 39.37,
		"miles":       1 / 63360,
		"attoparsecs": 1 / 1.215,
	},
	"miles": {
		"metres":      1609,
		"inches":      63360,
		"attoparsecs": 52155,
	},
	"attoparsecs": {
		"metres": 1 / 32.408,
		"inches": 1 / 1.215,
		"miles":  52155,
	},
	"kilograms": {
		"pounds":                 0.0,
		"ounces":                 0.0,
		"hogsheads of Beryllium": 0.0,
	},
	"pounds": {
		"kilograms":              0.0,
		"ounces":                 0.0,
		"hogsheads of Beryllium": 0.0,
	},
	"ounces": {
		"kilograms":              0.0,
		"pounds":                 0.0,
		"hogsheads of Beryllium": 0.0,
	},
	"hogsheads of Beryllium": {
		"kilograms": 0.0,
		"pounds":    0.0,
		"ounces":    0.0,
	},
}

var conversionPattern = regexp.MustCompile(`([^ ]*) (.*) to (.*)`)

var zPrint = fmt.Println

func main() {
	_, _ = zPrint("hello world")
}

func parseConversion(input string) (amount float64, fromUnit, toUnit string) {
	match := conversionPattern.FindStringSubmatch(input)
	amount, _ = strconv.ParseFloat(match[1], 64)
	return amount, match[2], match[3]
}

func convert(input string) (output string) {
	amount, fromUnit, toUnit := parseConversion(input)
	var result string
	conversion, exists := unitConversions[fromUnit][toUnit]
	if exists {
		result = fmt.Sprintf("is %f %s", amount*conversion, toUnit)
	} else {
		result = fmt.Sprintf("cannot be converted to %s", toUnit)
	}
	return fmt.Sprintf("%f %s %s", amount, fromUnit, result)
}
